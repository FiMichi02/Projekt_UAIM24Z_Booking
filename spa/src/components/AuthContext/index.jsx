import React, { createContext, useContext, useState, useEffect } from 'react';
import axios from 'axios';

// Tworzymy kontekst
const AuthContext = createContext();

// Hook, który będzie dostarczać dostęp do kontekstu
export const useAuth = () => {
    return useContext(AuthContext);
};

// Komponent, który będzie zarządzać stanem logowania
export const AuthProvider = ({ children, value = {} }) => {
    // Używamy wartości z `values` lub ustawiamy domyślne stany
    const [isLoggedIn, setIsLoggedIn] = useState(value.isLoggedIn ?? false);

    useEffect(() => {
        if (!value.skipCheck) {
            checkLoginStatus();
        }
    }, [value.skipCheck]);

    async function checkLoginStatus() {
        try {
            const response = await axios.get("http://172.20.0.40:5000/api/check-token", { withCredentials: true });
            setIsLoggedIn(response.data.isLoggedIn);
        } catch (error) {
            console.error("Error checking login status:", error);
            setIsLoggedIn(false);
        }
    }

    async function handleLogout() {
        try {
            await axios.post("http://172.20.0.40:5000/api/logout", {}, { withCredentials: true });
            setIsLoggedIn(false);
        } catch (error) {
            console.error("Error during logout:", error);
        }
    }

    return (
        <AuthContext.Provider
            value={{
                isLoggedIn,
                setIsLoggedIn,
                handleLogout: value.handleLogout || handleLogout,
                checkLoginStatus: value.checkLoginStatus || checkLoginStatus,
            }}
        >
            {children}
        </AuthContext.Provider>
    );
};
