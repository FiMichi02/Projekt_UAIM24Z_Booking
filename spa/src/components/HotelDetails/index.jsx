import React, {useEffect} from "react";
import './HotelDetails.css';
import {useLocation, useNavigate} from 'react-router-dom';

function HotelDetails() {
    const location = useLocation();
    const { hotel, detailsParams } = location.state || {};
    const navigate = useNavigate();

    useEffect(() => {
        console.log(detailsParams)
    });

    if (!hotel) {
        return <p style={{display:"flex", justifyContent:"center"}}>Hotel not found!</p>;
    }

    // Funkcja obsługująca kliknięcie przycisku rezerwacji
    const handleBookClick = (hotel, roomId) => {
        navigate("/new-booking", { state: { hotel, roomId, detailsParams } })
    };

    return (
        <div className="hotel-details-container">
            <h1>{hotel.Name}</h1>
            <div className="hotel-info">
                <p><strong>City:</strong> {hotel.CityName}</p>
                <p><strong>Country:</strong> {hotel.CountryName}</p>
            </div>

            <h2>Available Rooms</h2>
            <div className="rooms-container">
                {hotel.available_rooms.map(room => (
                    <div key={room.ID} className="room-card">
                        <p><strong>Room Size:</strong> {room.Size} people</p>
                        <p><strong>Price per Night:</strong> ${room.PricePerNight}</p>
                        <button
                            onClick={() => handleBookClick(hotel, room.ID)}
                            className="book-button"
                        >
                            Book
                        </button>
                    </div>
                ))}
            </div>
        </div>
    );
}

export default HotelDetails;
