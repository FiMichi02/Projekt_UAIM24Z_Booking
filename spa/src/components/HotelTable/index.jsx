import React from 'react';
import { Table } from 'antd';
import "./HotelTable.css";

const HotelTable = ({ hotels, onDetailsClick = f => f }) => {

    const columns = [
        {
            title: 'Name',
            dataIndex: 'Name',
            key: 'name',
            width: '25%',
        },
        {
            title: 'City',
            dataIndex: 'CityName',
            key: 'city',
            width: '25%',
        },
        {
            title: 'Country',
            dataIndex: 'CountryName',
            key: 'country',
            width: '25%',
        },
        {
            title: 'Book',
            key: 'book',
            width: '25%',
            render: (text, record) => (
                <div className="row-button-container">
                    <button
                        onClick={() => onDetailsClick(record.ID)}>
                        Book
                    </button>
                </div>
            ),
        },
    ];

    const expandedRowRender = (record) => {
        const roomColumns = [
            {
                title: 'Size',
                dataIndex: 'Size',
                key: 'size',
                width: '50%',
            },
            {
                title: 'Price Per Night',
                dataIndex: 'PricePerNight',
                key: 'pricePerNight',
                width: '50%',
                render: (text) => `$${text}`
            },
        ];

        return (
            <div className="ant-table-expanded-row-content">
                <Table
                    columns={roomColumns}
                    dataSource={record.available_rooms}
                    pagination={false}
                    rowKey="ID"
                />
            </div>
        );
    };

    return (
        <div className="table-container">
            <Table
                columns={columns}
                dataSource={hotels}
                expandable={{ expandedRowRender }}
                pagination={false}
                rowKey="ID"
            />
        </div>
    );
};

export default HotelTable;
