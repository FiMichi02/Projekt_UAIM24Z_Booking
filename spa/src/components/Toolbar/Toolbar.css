/* Toolbar.css */

/* Toolbar container */
.toolbar {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0 1.5rem;
    background-color: #b56576; /* Dirty pink background */
    color: #ffffff;
    font-family: Arial, sans-serif;
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.2);
    height: 50px; /* Fixed height */
    line-height: 50px; /* Ensure text elements align with the height */
}

/* Logo or Title */
.toolbar-logo {
    display: flex;
    align-items: center; /* Center the title vertically */
}

.toolbar-logo h1 {
    font-size: 1.5rem;
    margin: 0;
    color: #ffffff;
}

/* Navigation Links */
.toolbar-nav {
    display: flex;
    align-items: center; /* Center links vertically */
    gap: 1.5rem;
    font-weight: bold;
    font-size: 20px;
}

.toolbar-nav button {
    padding: 8px 16px !important; /* Override any inherited padding */
    font-size: 14px !important;    /* Override inherited font size */
    font-weight: 500;
    color: #b56576;
    background-color: #ffffff;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.2s ease, color 0.2s ease;
    margin: 0; /* Remove any inherited margin */
}

.toolbar-nav button:hover {
    background-color: #f5e1e1; /* Light pink */
    color: #a04d63;
}

/* Action Buttons (e.g., Login, Sign Up) */
.toolbar-actions {
    display: flex;
    align-items: center; /* Center buttons vertically */
    gap: 1rem;
}

.toolbar-actions button {
    padding: 8px 16px !important; /* Override any inherited padding */
    font-size: 14px !important;    /* Override inherited font size */
    font-weight: 500;
    color: #b56576;
    background-color: #ffffff;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.2s ease, color 0.2s ease;
    margin: 0; /* Remove any inherited margin */
}


.toolbar-actions button:hover {
    background-color: #f5e1e1; /* Light pink */
    color: #a04d63;
}

/* Responsive Design */
@media (max-width: 768px) {
    .toolbar-nav {
        display: none; /* Hide navigation on small screens */
    }
}
