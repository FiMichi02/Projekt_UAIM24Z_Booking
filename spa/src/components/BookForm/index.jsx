import '../SearchForm/SearchForm.css';
import axios from "axios";
import React, { useEffect, useState } from "react";
import {useLocation, useNavigate} from "react-router-dom";
import "./BookForm.css";
import {useAuth} from "../AuthContext";

function BookForm() {
    const location = useLocation()
    const { hotel, roomId, detailsParams } = location.state || {};
    const { handleLogout } = useAuth();
    const navigate = useNavigate();

    const [localDetailsParams, setLocalDetailsParams] = useState(detailsParams || {});

    useEffect(() => {
        if (detailsParams) {
            setLocalDetailsParams(detailsParams);
        }
    }, [detailsParams]);

    const [localHotel, setLocalHotel] = useState(hotel || {})

    useEffect(() => {
        if (hotel) {
            setLocalHotel(hotel);
        }
    }, [hotel]);

    const [selectedRoomId, setSelectedRoomId] = useState(roomId);

    useEffect(() => {
        if (localHotel.available_rooms?.length > 0 && localDetailsParams.size) {
            const matchingRoom = localHotel.available_rooms.find(
                (room) => (room.Size === parseInt(localDetailsParams.size))
            );

            const currentRoom = localHotel.available_rooms.find(
                (room) => (room.ID === selectedRoomId)
            );

            if (matchingRoom && currentRoom.Size < parseInt(localDetailsParams.size)) {
                setSelectedRoomId(matchingRoom.ID);
            }
        }
    }, [localDetailsParams.size, localHotel, selectedRoomId]);

    const handleInputChange = (e) => {
        const { name, value } = e.target;
        setLocalDetailsParams((prev) => ({
            ...prev,
            [name]: value
        }));
    };

    // Aktualizacja dostępnych pokoi po zmianie wartości formularza
    useEffect(() => {
        axios
            .post("http://172.20.0.40:5000/api/get-hotels", localDetailsParams)
            .then((response) => {
                setLocalHotel(response.data[0]);
            })
            .catch((error) => {
                console.error("Error fetching hotels:", error);
            });
    }, [localDetailsParams]);


    function book(event) {
        event.preventDefault(); // Zapobiega przeładowaniu strony
        if (!selectedRoomId) {
            alert("Please select a room before booking!");
            return;
        }
        const data = {
            room_id: selectedRoomId,
            hotel_id: localHotel.ID,
            from: localDetailsParams.from,
            to: localDetailsParams.to
        }
        axios.post("http://172.20.0.40:5000/api/make-reservation", data, {withCredentials: true})
            .then(response => {
                alert(response.data.Success)
                navigate("/profile")
            })
            .catch(error => {
                if(error.response.status === 401) {
                    alert("Error: Session probably expired, please log in again and try again")
                    handleLogout()
                    navigate("/login")
                }
            })
    }


    function handleRoomChange(newRoomId) {
        setSelectedRoomId(newRoomId)
        console.log(localDetailsParams)
    }

    return (
        <div className={"search-container"}>
            <form onSubmit={book}>
                <h1>Please, fill the form to complete the booking </h1>
                <h2>{localHotel.Name}</h2>
                <h2>{localHotel.CountryName}, {localHotel.CityName}</h2>
                <div className={"search-adults-children"}>
                    <div className={"input-group"}>
                        <label htmlFor="people">People:</label>
                        <input id="people" name="size" value={localDetailsParams.size} type={"number"} placeholder={"No of people"} onChange={handleInputChange} required/>
                    </div>
                </div>
                <div className={"search-date"}>
                    <div className={"input-group"}>
                        <label htmlFor="from">From:</label>
                        <input name="from" value={localDetailsParams.from} placeholder={"From..."} type={"date"} onChange={handleInputChange} required/>
                    </div>
                    <div className={"input-group"}>
                        <label htmlFor="to">To:</label>
                        <input name="to" value={localDetailsParams.to} placeholder={"...To"} type={"date"} onChange={handleInputChange} required/>
                    </div>
                </div>
                <h2>Available Rooms</h2>
                <div className="rooms-container">
                    {localHotel.available_rooms.map(room => (
                        <div key={room.ID} className="room-card">
                            <p><strong>Room Size:</strong> {room.Size} people</p>
                            <p><strong>Price per Night:</strong> ${room.PricePerNight}</p>
                            <button
                                type="button"
                                onClick={() => handleRoomChange(room.ID)}
                                className={`new-room-button ${selectedRoomId === room.ID ? "selected" : ""}`}
                            >
                                {selectedRoomId === room.ID ? "Selected" : "Book"}
                            </button>
                        </div>
                    ))}
                </div>
                <button type={"submit"}>Book!</button>
            </form>
        </div>
    )
}

export default BookForm;
