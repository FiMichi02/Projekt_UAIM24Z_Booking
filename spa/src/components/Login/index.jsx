import LoginForm from './LoginForm'
import React from 'react'
import './Login.css'
import axios from "axios";
import { useNavigate } from "react-router-dom";
import { useAuth } from '../AuthContext'; // importuj kontekst

function Login() {
    const navigate = useNavigate();
    const { setIsLoggedIn } = useAuth(); // dostęp do funkcji zmiany stanu logowania

    return (
        <div className={"login-container"}>
            <LoginForm getCreds={
                function (email, password) {
                    const data = {
                        email: email,
                        password: password
                    }

                    axios.post("http://172.20.0.40:5000/api/login", data, { withCredentials: true })
                        .then(response => {
                            console.log('Success:', response.data);
                            alert("Logged in successfully!");
                            setIsLoggedIn(true); // Zaktualizuj stan logowania
                            navigate("/"); // Przekieruj na stronę główną
                        })
                        .catch(error => {
                            alert("Email or Password incorrect!");
                        });
                }
            } />
        </div>
    );
}

export default Login;
