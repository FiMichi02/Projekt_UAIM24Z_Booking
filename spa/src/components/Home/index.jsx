
import SearchForm from "../SearchForm";
import Search from "../Search";
import {useEffect, useState} from "react";
import HotelDetails from "../HotelDetails";
import {Route, Routes, useNavigate} from "react-router-dom";

function Home() {

    const navigate = useNavigate()
    const [searchParams, setSearchParams] = useState({size: 1});

    // Funkcja do aktualizacji stanu, wywoÅ‚ywana przez SearchForm
    function handleSearch(name, size, from, to, country, city){
        setSearchParams({
            size: size, city: city, country: country,
            from: from, name: name, to: to
        });
        navigate("/search")
    }

    useEffect(() => {
        console.log(searchParams);
    }, [searchParams]);



    return(
        <div>
            <h1>Home Page</h1>
            <SearchForm searchPlace={
                handleSearch
            }></SearchForm>
            <Routes>
                <Route path="/search" element={<Search searchParams={searchParams}/>}/>
                <Route path="/hotel-details" element={<HotelDetails hotel={{id:1}}/>}/>
            </Routes>
        </div>
    )
}

export default Home;