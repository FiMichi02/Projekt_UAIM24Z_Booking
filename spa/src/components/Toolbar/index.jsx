import React from 'react';
import { useNavigate } from "react-router-dom";
import { useAuth } from '../AuthContext'; // importuj kontekst
import './Toolbar.css';

function Toolbar() {
    const navigate = useNavigate();
    const { isLoggedIn, handleLogout } = useAuth(); // pobieramy isLoggedIn z kontekstu

    function handleNavigate(path) {
        navigate(path);
    }

    function logout() {
        handleLogout()
        alert("Logged out!")
        navigate("/")
    }

    return (
        <div className="toolbar">
            <div className="toolbar-logo">
                <h1>Hotel Booking</h1>
            </div>
            <div className="toolbar-nav">
                <button onClick={() => handleNavigate("/")}>Home</button>
            </div>
            <div className="toolbar-actions">
                {isLoggedIn ? (
                    <>
                        <button onClick={() => handleNavigate("/profile")}>Profile</button>
                        <button onClick={() => logout()}>Logout</button>
                    </>
                ) : (
                    <>
                        <button onClick={() => handleNavigate("/login")}>Login</button>
                        <button onClick={() => handleNavigate("/signin")}>Sign Up</button>
                    </>
                )}
            </div>
        </div>
    );
}

export default Toolbar;
