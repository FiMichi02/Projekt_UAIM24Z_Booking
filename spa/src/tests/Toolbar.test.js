import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext, AuthProvider } from '../components/AuthContext';
import Toolbar from '../components/Toolbar';
import axios from 'axios';

// Mockowanie axios
jest.mock('axios');

describe('Toolbar component', () => {
    test('renders login and sign-up buttons when not logged in', async () => {
        axios.get.mockResolvedValueOnce({ data: { isLoggedIn: false } });

        render(
            <AuthProvider>
                <MemoryRouter future={{
                    v7_startTransition: true,
                    v7_relativeSplatPath: true,
                }}>
                    <Toolbar />
                </MemoryRouter>
            </AuthProvider>
        );

        await screen.findByText(/login/i);
        expect(screen.getByText(/login/i)).toBeInTheDocument();
        expect(screen.getByText(/sign up/i)).toBeInTheDocument();
    });

    test('renders profile and logout buttons when logged in', async () => {
        axios.get.mockResolvedValueOnce({ data: { isLoggedIn: true } });

        render(
            <AuthProvider>
                <MemoryRouter future={{
                    v7_startTransition: true,
                    v7_relativeSplatPath: true,
                }}>
                    <Toolbar />
                </MemoryRouter>
            </AuthProvider>
        );

        await screen.findByText(/profile/i);
        expect(screen.getByText(/profile/i)).toBeInTheDocument();
        expect(screen.getByText(/logout/i)).toBeInTheDocument();
    });

    test('calls logout and navigates to home page when Logout button is clicked', async () => {
        const mockHandleLogout = jest.fn();
        axios.get.mockResolvedValueOnce({ data: { isLoggedIn: true } });
        axios.post.mockResolvedValueOnce({});

        global.alert = jest.fn();

        render(
            <AuthProvider value={{ isLoggedIn: true, handleLogout: mockHandleLogout }}>
                <MemoryRouter future={{
                    v7_startTransition: true,
                    v7_relativeSplatPath: true,
                }}>
                    <Toolbar />
                </MemoryRouter>
            </AuthProvider>
        );

        fireEvent.click(screen.getByText(/logout/i));

        expect(mockHandleLogout).toHaveBeenCalled();
    });
});
