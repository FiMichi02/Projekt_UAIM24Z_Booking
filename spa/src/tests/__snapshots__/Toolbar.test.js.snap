// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Component renders and matches snapshot 1`] = `
<div>
  <div
    class="toolbar"
  >
    <div
      class="toolbar-logo"
    >
      <h1>
        Hotel Booking
      </h1>
    </div>
    <div
      class="toolbar-nav"
    >
      <button>
        Home
      </button>
    </div>
    <div
      class="toolbar-actions"
    >
      <button>
        Login
      </button>
      <button>
        Sign Up
      </button>
    </div>
  </div>
</div>
`;
